keps/sig-node/4369-allow-special-characters-environment-variable/README.md

```diff
@@ -0,0 +1,319 @@
+
+### Non-Goals
+
+## Proposal
+
+* Implements relaxed validation at the top level validation method when validating API create requests.
+* Document rules for setting environment variables.
+
+### User Stories (Optional)
+
+#### Story 1
+
+I am a .NET Core development engineer, .Net Core applications are using ":" when working with application settings loaded from appsettings.json file. When running .net core app in containers typically overwrite this settings by specifying environmental variable.
+such as: 
+`"Logging": { "IncludeScopes": false, "LogLevel": { "Default": "Warning" } }`    
+override like this `-e Logging:LogLevel:Default=Debug`    
+
+### Risks and Mitigations
+
+Relaxed validation can break upgrade and rollback scenarios, but our use of feature gate to control whether it's enabled or not will make it a manageable risk, with the user having the autonomy to choose whether or not to enable it.
+
+## Design Details
+
+- A feature gate name `RelaxedEnvironmentVariableValidation` controlling the loosening of the envvar name validation, initially in alpha state and defaulting to false
+- Two sets of validation logic for envvar names:
+  - Current validation
+  - Relaxed validation
+- Everywhere we validate envvar names in API objects, plumbing a parameter whether we want the strict or relaxed validation
+  - At the top level validation method when validating API create requests, use the strict validation if the feature gate is off
+  - At the top level validation method when validating API update requests, use the strict validation if the feature gate is off and the old object passes strict envvar name validation
+
+### Test Plan
+
+[x] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+##### Unit tests
+
+Currently coverages:
+
+* pkg/apis/core/validation/validation_test.go: `2023-12-21` - `83.9%`
+* pkg/kubelet/kubelet_pods_test.go: `2023-12-21` - `67.2%`
+* staging/src/k8s.io/apimachinery/pkg/util/validation/validation_test.go: `2023-12-21` - `94.8%`
+
+These tests will be added:
+
+* New tests will be added to ensure environment variable fields can be correctly validated `pkg/apis/core/validation/validation_test.go`
+* Add a new test that sets special character environment variables for pods in a given namespace `pkg/kubelet/kubelet_pods_test.go`
+* A new test will be added to ensure that the environment variable name field is valid `staging/src/k8s.io/apimachinery/pkg/util/validation/validation_test.go`
+
+##### Integration tests
+
+- N/A
+
+##### e2e tests
+
+* Add a test to `test/e2e/common/configmap.go` to test that the special characters in configmap are consumed by the environment variable.
+
+* Add a test to `test/e2e/common/secret.go` to test that the special characters in secret are consumed by the environment variable.
```

Comment:
This KEP *does not* cover changing secrets or configmaps to allow a broader spec for keys.  If we want to do that, it's a different KEP.

Reply:
not allowing broader keys, but allowing EnvFrom to consume the full range of existing configmap and secret keys, right? today, the keyspace for configmaps/secrets is more broad than envvar names

Reply:
ahh, I get you.

`IsConfigMapKey()` checks:
- len <= 253; and
- regex `^[-._a-zA-Z0-9]+$`; and
- not "." or ".."; and
- does not begin with ".."

`IsEnvVarName()` checks:
- regex `^[-._a-zA-Z][-._a-zA-Z0-9]*$`
- not "." or ".."; and
- does not begin with ".."

The main difference being that CM keys can start with a digit and env can't (or I can't read regex).

`kubelet.makeEnvironmentVariables()` filters CM keys to those that pass `IsEnvVarName()`.

So you are right that CM keys which used to be disallowed will be allowed.

Reply:
> That wasn't really discussed in the KEP. Is allowing that also controlled by the feature gate?

I will remove this part and add them to `Non-Goals`.

> So you are right that CM keys which used to be disallowed will be allowed.

`IsConfigMapKey()` is used in the `validation` of secret and configmap to verify the character range of the key, so we didn't make destructive changes to them(we will not change `IsConfigMapKey()`). The configmap key that couldn't be used before is still unusable now (because it can't be created)

Reply:
> The configmap key that couldn't be used before is still unusable now (because it can't be created)

I know we're not changing IsConfigMapKey(), but today, IsConfigMapKey() allows things into configmaps which are discarded when using them as envvars with `EnvFrom`.

I was asking what this feature would change (if anything) if a user had a configmap with a key like `0123` that didn't pass C_IDENTIFIER validation. That is allowed in a configmap today and is discarded when mapping the configmap keys to envvars.

Reply:
Thank you for the reminder. I overlooked the details here earlier, and you're right. It's an omission in the KEP. I believe using ConfigMap and Secret in EnvFrom should be part of this KEP, and I'll make sure to add it later.